src/include/units/zero_rep.h

```diff
@@ -0,0 +1,149 @@
+// The MIT License (MIT)
+//
+// Copyright (c) 2020 Yves Delley
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+#pragma once
+
+#include <units/one_rep.h>
+#include <units/bits/basic_concepts.h>
+#include <units/customization_points.h>
+#include <type_traits>
+
+namespace units {
+
+struct invalid_zero_rep {};
+
+/**
+ * @brief A representation type to be used for unit constants
+ *
+ * This representation type is intended to be used in the unit constants definition:
+ *
+ * @code{.cpp}
+ * namespace unit_constants {
+ * 
+ * inline constexpr auto zero_celsius_point = thermodynamic_temperature_point<degree_celsius, zero_rep, celsius_temperature_origin>{};
+ *
+ * }
+ * @endcode
+ * 
+ * Unit constants simplify quantity point creation:
+ * 
+ * @code{.cpp}
+ * using namespace units::physical::si::unit_constants;
+ *
+ * auto d = 25 * deg_C + zero_celsius_point;
+ * @endcode
+ *
+ */
+struct zero_rep {
```

Comment:
The whole purpose of introduction of `one_rep` was to limit invalid operations so it cannot be used in a wrong way. `zero_rep` seems to strive to be fully arithmetic which actually does not help much over any other regular arithmetic type. I would limit its usage only to addition (and possibly subtraction, although I am not sure if it is even needed here).

Can we come up with any sane reason to multiply or divide it?

Reply:
In other words, I understand why you wanted for `zero_rep` to behave like this and why did you complain about `one_rep` below. But it is not about maths here but interfaces. Maybe just the types are named in the wrong way and we should rename them to be less confusing.

Reply:
The most basic reason to allow multiplication and division on any of those types is to conform to the concept of the QuantityValue as it is intended. If we don't do that, we hinder separation of concerns in the library design. The "representation slot" asks for a numeric type, and we are just providing special-purpose compile-time values instead of runtime values here. If this weren't such a low-level library with minimal dependencies, I would just use `boost::hana::int_c<0>`. IMHO, limiting operations should be done on the actual interface type instead, so instead of using `quantity` for the unit constants, declare `unit_constant` and similarly for the origin constants.

This solution is of course more pragmatic. The `zero_rep` needs multiplication however: The `zp_degree_celsius` is a constant quantity point that can be mathematically expanded to roughly `(zero * degree_celsius{} + celsius_temperature_origin{})`. The expression `zp_degree_celsius + 25 * deg_C` then roughly expands to `(zero * degree_celsius{} + celsius_temperature_origin{}) + 25 * (one * degree_celsius{})` which resolves to `(zero * degree_celsius{} + 25 * degree_celsius{}) + celsius_temperature_origin{}`. Pretty sure there's sums and multiplications involved, can't say if as lhs or rhs. Keeping track what what exact operation of the representation type will actually be invoked is hard, but as long as those types follow their concept, we know what will come out.

Reply:
However, the `zero_rep` here indeed is just an optimisation, guaranteeing compile-time evaluation of those operations. If the implied complexity is higher than the value of the performance gain, we can of course just stick with an `int`.

Reply:
Actually, make this an `int8_t`. I am using this library in embedded systems, where I want to tightly control the representation of my quantity points and would loathe automatic widening to int.

Reply:
> limiting operations should be done on the actual interface type instead, so instead of using quantity for the unit constants, declare unit_constant and similarly for the origin constants.

Actually, this is what was suggested by the ISO C++ Committee and I wanted to go in this direction. Alternatively, Walter Brown was always arguing that `metre` is just a constant quantity of length with the value `1`.

In the meantime, @johelegp came with an out-of-the-box solution of `one_rep` which did the trick in an interesting way that I was initially not thinking about. Such a solution was more compatible with what Walter suggested and I thought it is nice at that time.

Reply:
> Actually, make this an int8_t

`int8_t` is (or may not be) portable so it should not be used inside the C++ Standard Library if only possible.

Reply:
Then, I'd like to stick with a compiler-time number, such as a `zero_rep`. Because otherwise, it would cause widening of the representation of whatever quantity is put in front of the origin constant.

Reply:
You can use `int_least8_t` which is portable ;-)

Reply:
But that wouldn't be as portable as using the `Rep` of the other "hand side".

Reply:
What are the concerns of leaving this at `zero_rep` for now?

Reply:
I like `zero_rep` and `one_rep` should probably be made consistent. Then they will be really nice mathematical types and maybe someone will use them :wink:. With this, I think such a solution will not provide a strong easy to use and hard to abuse interface for `quantity` and `quantity_point` creation and we should not use those types for this purpose in such a case.

We want a type-safe interface. Actually, during the discussion in the ISO C++ Committee I got a suggestion that instead of the explicit constructor taking the value the only way to create the quantity should be to multiply a value with some `magic_type` that caries up the information about dimension and unit. Something like this:

```cpp
magic_type<si::dim_length, si::metre> m{};
si::length<si::metre> d = 123 * m;
```

The above should be the only operation supported for magic. Specifically, `magic * value`, division, additions, subtractions, modulus should always result in a compile-time error. There are people in the Committee that are already concerned with potential issues of explicit constructor's usage. I think that `zero_rep` as proposed will not make them happy (although some of them have a really strong mathematical background and are members of SG4).

Reply:
Ok, then I'll create that `magic_origin_point` type and let someone else handle the `magic_unit_type`. Don't worry, I won't name it "magic".

Reply:
On which side of the "+" should the origin stand? I find `zeroth_deg_celsius + 3*deg_C` slightly better than `3*deg_C + zeroth_deg_celsius`, but maybe both could be allowed. Also, the purists from the committee would therefore probably disallow `zeroth_deg_celsius - 3*deg_C` and prefer `zeroth_deg_celsius + -3*deg_C`, right?

To remove that perceived redundancy, we could still consider `3 | deg_C` or any other op and/or a different name. However, none of them has the mathematical accuracy of the above.

Reply:
I think _`zeroth`_ `+`/`-` `q` is fine.

Reply:
You can achieve that with `zero_rep`.

Reply:
> You can achieve that with zero_rep.

Yeah, this is why I liked your solution for `one_rep`. But maybe both of them should be renamed to sound less mathematical so no one will complain that they do not model "zero" and "one" properly.

Reply:
How about `unit_constant_rep` and `point_origin_rep`?

Reply:
Ok, I'll therefore rename to `point_origin_rep` and leave it at that for now.

Reply:
Are you going to leave all of the "unsafe" operators of `zero_rep`? The main reason to rename those types is to enforce compile-time errors on invalid usage while not making any claims about its mathematical correctness.